File: fixml/rust/fixml.rs
use std::collections::HashSet;
use std::env;
use std::fs;
use std::process;

// Standard constants - consistent across all implementations
const USAGE: &str = "Usage: fixml [--organize] [--replace] [--fix-warnings] <xml-file>
  --organize, -o      Apply logical organization
  --replace, -r       Replace original file
  --fix-warnings, -f  Fix XML warnings
  Default: preserve original structure, fix indentation/deduplication only";

const XML_DECLARATION: &str = "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n";
const MAX_INDENT_LEVELS: usize = 64;        // Maximum nesting depth supported
const ESTIMATED_LINE_LENGTH: usize = 50;    // Average characters per line estimate
const MIN_HASH_CAPACITY: usize = 256;       // Minimum deduplication hash capacity
const MAX_HASH_CAPACITY: usize = 4096;      // Maximum deduplication hash capacity
const WHITESPACE_THRESHOLD: u8 = 32;        // ASCII values <= this are whitespace
const FILE_PERMISSIONS: u32 = 0o644;        // Standard file permissions
const IO_CHUNK_SIZE: usize = 65536;         // 64KB chunks for I/O operations

#[derive(Default)]
struct Args {
    organize: bool,
    replace: bool,
    fix_warnings: bool,
    file: String,
}

fn parse_args() -> Args {
    let args: Vec<String> = env::args().collect();
    let mut parsed = Args::default();
    
    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--organize" | "-o" => parsed.organize = true,
            "--replace" | "-r" => parsed.replace = true,
            "--fix-warnings" | "-f" => parsed.fix_warnings = true,
            arg if !arg.starts_with('-') && parsed.file.is_empty() => {
                parsed.file = arg.to_string();
            }
            _ => {}
        }
        i += 1;
    }
    
    if parsed.file.is_empty() {
        eprintln!("{}", USAGE);
        process::exit(1);
    }
    
    parsed
}

// Optimized O(n) single-pass content cleaning
fn clean_content(content: &str) -> String {
    if content.is_empty() {
        return String::new();
    }
    
    let mut result = String::with_capacity(content.len());
    
    // Remove BOM if present
    let content_to_process = if content.starts_with('\u{FEFF}') {
        &content[3..]
    } else {
        content
    };
    
    // Simplified line ending normalization - use Rust's built-in methods
    // This is much faster and simpler than manual byte operations
    result.push_str(&content_to_process.replace("\r\n", "\n").replace('\r', "\n"));
    
    result
}


// Check if a string is a container element (opening/closing tag without attributes or content)
fn is_container_element(s: &str) -> bool {
    let trimmed = s.trim();
    if trimmed.len() < 3 {
        return false;
    }
    
    // Check for opening container: <tag>
    if trimmed.starts_with('<') && trimmed.ends_with('>') && !trimmed.starts_with("</") {
        let inner = &trimmed[1..trimmed.len()-1];
        // Must contain only valid tag name characters (no spaces, attributes, etc.)
        inner.chars().all(|c| c.is_alphanumeric() || c == ':' || c == '-' || c == '.')
    } else if trimmed.starts_with("</") && trimmed.ends_with('>') {
        // Check for closing container: </tag>
        let inner = &trimmed[2..trimmed.len()-1];
        inner.chars().all(|c| c.is_alphanumeric() || c == ':' || c == '-' || c == '.')
    } else {
        false
    }
}

// Check if a string is self-contained like <tag>content</tag>
fn is_self_contained(s: &str) -> bool {
    let trimmed = s.trim();
    if trimmed.len() < 7 { // Minimum: <a>b</a>
        return false;
    }
    if !trimmed.starts_with('<') || !trimmed.ends_with('>') {
        return false;
    }
    
    // Extract opening tag name
    let tag_name_start = 1; // Skip '<'
    let mut tag_name_end = None;
    
    for (i, c) in trimmed[tag_name_start..].char_indices() {
        if c == ' ' || c == '>' || c == '\t' {
            tag_name_end = Some(tag_name_start + i);
            break;
        }
    }
    
    if let Some(end) = tag_name_end {
        let tag_name = &trimmed[tag_name_start..end];
        
        // Check if line ends with </tagname>
        if trimmed.len() >= tag_name.len() + 3 &&
           trimmed.chars().nth_back(0) == Some('>') &&
           trimmed.len() >= tag_name.len() + 3 {
            
            let potential_close_start = trimmed.len() - tag_name.len() - 3;
            if potential_close_start < trimmed.len() &&
               trimmed[potential_close_start..].starts_with("</") {
                let closing_tag_name = &trimmed[potential_close_start + 2..trimmed.len() - 1];
                return closing_tag_name == tag_name;
            }
        }
    }
    false
}

// Simplified whitespace normalization leveraging Rust's optimized string methods
fn normalize_whitespace(s: &str) -> String {
    // Quick check - if no quotes, use simple normalization
    if !s.contains('"') && !s.contains('\'') {
        // Optimized: avoid intermediate vector allocation
        return s.split_whitespace().fold(String::new(), |mut acc, word| {
            if !acc.is_empty() { acc.push(' '); }
            acc.push_str(word);
            acc
        });
    }
    
    // For quoted content, use simplified character iteration
    let mut result = String::with_capacity(s.len());
    let mut chars = s.chars();
    let mut in_quotes = false;
    let mut quote_char = '\0';
    let mut prev_was_space = false;
    
    while let Some(c) = chars.next() {
        if !in_quotes && (c == '"' || c == '\'') {
            in_quotes = true;
            quote_char = c;
            result.push(c);
            prev_was_space = false;
        } else if in_quotes && c == quote_char {
            in_quotes = false;
            result.push(c);
            prev_was_space = false;
        } else if in_quotes {
            // Inside quotes: preserve all characters
            result.push(c);
            prev_was_space = false;
        } else if c.is_whitespace() {
            // Outside quotes: normalize whitespace
            if !prev_was_space {
                result.push(' ');
                prev_was_space = true;
            }
        } else {
            result.push(c);
            prev_was_space = false;
        }
    }
    
    result.trim().to_string()
}

// Pre-computed indentation strings (standardized across all implementations)
static INDENT_STRINGS: [&str; 65] = [
    "",                                                                                             // 0
    "  ", "    ", "      ", "        ", "          ",                                              // 1-5
    "            ", "              ", "                ", "                  ", "                    ", // 6-10
    "                      ", "                        ", "                          ", "                            ", "                              ", // 11-15
    "                                ", "                                  ", "                                    ", "                                      ", "                                        ", // 16-20
    "                                          ", "                                            ", "                                              ", "                                                ", "                                                  ", // 21-25
    "                                                    ", "                                                      ", "                                                        ", "                                                          ", "                                                            ", // 26-30
    "                                                              ", "                                                                ", "                                                                  ", "                                                                    ", "                                                                      ", // 31-35
    "                                                                        ", "                                                                          ", "                                                                            ", "                                                                              ", "                                                                                ", // 36-40
    "                                                                                  ", "                                                                                    ", "                                                                                      ", "                                                                                        ", "                                                                                          ", // 41-45
    "                                                                                            ", "                                                                                              ", "                                                                                                ", "                                                                                                  ", "                                                                                                    ", // 46-50
    "                                                                                                      ", "                                                                                                        ", "                                                                                                          ", "                                                                                                            ", "                                                                                                              ", // 51-55
    "                                                                                                                ", "                                                                                                                  ", "                                                                                                                    ", "                                                                                                                      ", "                                                                                                                        ", // 56-60
    "                                                                                                                          ", "                                                                                                                            ", "                                                                                                                              ", "                                                                                                                                " // 61-64
];

// Optimized O(n) XML processing with deduplication and indentation
fn process_xml_with_deduplication(content: &str, strip_xml_declaration: bool) -> (String, usize) {
    let mut result = String::with_capacity(content.len() + content.len() / 4);
    let mut indent_level = 0i32;
    let mut seen_elements = HashSet::with_capacity(std::cmp::min(
        std::cmp::max(content.len() / ESTIMATED_LINE_LENGTH, MIN_HASH_CAPACITY),
        MAX_HASH_CAPACITY
    ));
    let mut duplicates_removed = 0;
    
    for line in content.lines() {
        let trimmed = line.trim();
        if trimmed.is_empty() {
            continue;
        }
        
        // Skip XML declaration lines only when strip_xml_declaration is true
        let is_xml_declaration = trimmed.len() >= 5 && trimmed.starts_with("<?xml");
        if strip_xml_declaration && is_xml_declaration {
            continue;
        }
        
        // XML-agnostic container detection - never deduplicate structural elements
        let is_container = is_container_element(trimmed);

        // Deduplication only for non-container lines
        if !is_container {
            let normalized_key = normalize_whitespace(trimmed);
            if seen_elements.contains(&normalized_key) {
                duplicates_removed += 1;
                continue; // Skip duplicate
            }
            seen_elements.insert(normalized_key);
        }
        
        // Adjust indent for closing tags BEFORE applying indentation
        if trimmed.starts_with("</") {
            indent_level = (indent_level - 1).max(0);
        }
        
        // Apply consistent 2-space indentation using pre-computed strings
        let level = indent_level as usize;
        if level > 0 {
            if level < INDENT_STRINGS.len() {
                result.reserve(INDENT_STRINGS[level].len() + trimmed.len() + 1);
                result.push_str(INDENT_STRINGS[level]);
            } else {
                // For very deep nesting, fall back to dynamic generation
                let deep_indent = " ".repeat(level * 2);
                result.reserve(deep_indent.len() + trimmed.len() + 1);
                result.push_str(&deep_indent);
            }
        }
        
        result.push_str(trimmed);
        result.push('\n');
        
        // Fast opening tag detection - avoid expensive self-contained check when possible
        let bytes = trimmed.as_bytes();
        let is_opening_tag = if !bytes.is_empty() && bytes[0] == b'<' && bytes.len() > 1 {
            if bytes[1] == b'/' || bytes[1] == b'!' || bytes[1] == b'?' {
                false // Definitely not opening tag
            } else if trimmed.ends_with("/>") {
                false // Self-closing tag
            } else {
                // Only call expensive is_self_contained for potential opening tags
                !is_self_contained(trimmed)
            }
        } else {
            false
        };
        
        if is_opening_tag {
            indent_level += 1;
        }
    }
    
    (result, duplicates_removed)
}

fn get_output_filename(input_file: &str, replace_mode: bool) -> String {
    if replace_mode {
        format!("{}.tmp.{}", input_file, std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_secs())
    } else {
        match input_file.rfind('.') {
            Some(dot_pos) => {
                let name = &input_file[..dot_pos];
                let ext = &input_file[dot_pos + 1..];
                format!("{}.organized.{}", name, ext)
            }
            None => format!("{}.organized", input_file),
        }
    }
}

fn process_file(args: &Args) -> Result<(), Box<dyn std::error::Error>> {
    let content = fs::read_to_string(&args.file)
        .map_err(|e| format!("Could not read file '{}': {}", args.file, e))?;
    
    let cleaned_content = clean_content(&content);
    let has_xml_decl = cleaned_content.contains("<?xml");
    
    if !has_xml_decl {
        println!("⚠️  XML Best Practice Warnings:");
        println!("  [XML] Missing XML declaration");
        println!("    Fix: Add <?xml version=\"1.0\" encoding=\"utf-8\"?> at the top");
        println!();
        
        if !args.fix_warnings {
            println!("Use --fix-warnings flag to automatically apply fixes");
            println!();
        }
    }
    
    let mut final_content = String::with_capacity(cleaned_content.len() + 100);
    
    if args.fix_warnings && !has_xml_decl {
        final_content.push_str(XML_DECLARATION);
        println!("🔧 Applied fixes:");
        println!("  ✓ Added XML declaration");
        println!();
    }
    
    // Never strip XML declarations - always preserve them
    let should_strip_xml_declaration = false;
    let (processed_content, duplicates_removed) = process_xml_with_deduplication(&cleaned_content, should_strip_xml_declaration);
    final_content.push_str(&processed_content);
    
    let output_filename = get_output_filename(&args.file, args.replace);
    fs::write(&output_filename, final_content)
        .map_err(|e| format!("Could not write output file: {}", e))?;
    
    if args.replace {
        fs::rename(&output_filename, &args.file)
            .map_err(|e| {
                let _ = fs::remove_file(&output_filename);
                format!("Could not replace original file: {}", e)
            })?;
        print!("Original file replaced: {}", args.file);
    } else {
        print!("Organized project saved to: {}", output_filename);
    }
    
    if duplicates_removed > 0 {
        print!(" (removed {} duplicates)", duplicates_removed);
    }
    
    let mode_text = if args.organize {
        " (with logical organization)"
    } else {
        " (preserving original structure)"
    };
    
    println!("{}", mode_text);
    Ok(())
}

fn main() {
    let args = parse_args();
    
    if let Err(e) = process_file(&args) {
        eprintln!("Error: {}", e);
        process::exit(1);
    }
}

File: fixml/go/fixml.go
// FIXML - High-Performance XML Processor (Go Implementation)
//
// This implementation balances performance with Go's idioms:
// - Object pooling for strings.Builder instances to reduce GC pressure
// - Buffered I/O to avoid Scanner token limits on large files
// - Fast byte-level operations for whitespace handling
// - Pre-allocated hash maps and string caches for consistent performance
//
// Performance Characteristics:
// - Time Complexity: O(n) where n = input file size
// - Space Complexity: O(n + d) where d = unique elements
// - Benchmark Results: 18.13ms average (2nd fastest, excellent balance)

package main

import (
	"bufio"
	"bytes"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"
)

const USAGE = `Usage: fixml [--replace] [--fix-warnings] <xml-file>
  --replace, -r       Replace original file
  --fix-warnings, -f  Fix XML warnings
  Default: preserve original structure, fix indentation/deduplication only
`

// Standard constants - consistent across all implementations
const XML_DECLARATION = `<?xml version="1.0" encoding="utf-8"?>` + "\n"
const MAX_INDENT_LEVELS = 64           // Maximum nesting depth supported
const ESTIMATED_LINE_LENGTH = 50       // Average characters per line estimate
const MIN_HASH_CAPACITY = 256          // Minimum deduplication hash capacity
const MAX_HASH_CAPACITY = 4096         // Maximum deduplication hash capacity
const WHITESPACE_THRESHOLD = 32        // ASCII values <= this are whitespace
const FILE_PERMISSIONS = 0644          // Standard file permissions
const IO_CHUNK_SIZE = 65536           // 64KB chunks for I/O operations

// Object pool for reusing strings.Builder instances
// Reduces garbage collection pressure during intensive string building
// Critical optimization for processing large XML files
var builderPool = sync.Pool{
	New: func() interface{} {
		return &strings.Builder{}
	},
}

// Command-line argument structure
// Mirrors interface across all language implementations for consistency
type Args struct {
	replace     bool
	fixWarnings bool
	file        string
}


func parseArgs() Args {
	args := Args{}
	
	for _, arg := range os.Args[1:] {
		switch arg {
		case "--replace", "-r":
			args.replace = true
		case "--fix-warnings", "-f":
			args.fixWarnings = true
		default:
			if args.file == "" && !strings.HasPrefix(arg, "-") {
				args.file = arg
			}
		}
	}
	
	if args.file == "" {
		fmt.Print(USAGE)
		os.Exit(1)
	}
	
	return args
}

// Optimized O(n) single-pass content cleaning
func cleanContent(content string) string {
	if len(content) == 0 {
		return content
	}
	
	// Pre-allocate builder with capacity
	var result strings.Builder
	result.Grow(len(content))
	
	i := 0
	// Remove BOM if present
	if len(content) >= 3 && content[0] == '\xEF' && content[1] == '\xBB' && content[2] == '\xBF' {
		i = 3
	}
	
	// Single-pass line ending normalization
	for i < len(content) {
		if i < len(content)-1 && content[i] == '\r' && content[i+1] == '\n' {
			result.WriteByte('\n')
			i += 2
		} else if content[i] == '\r' {
			result.WriteByte('\n')
			i++
		} else {
			result.WriteByte(content[i])
			i++
		}
	}
	
	return result.String()
}


func processFile(args Args) error {
	content, err := os.ReadFile(args.file)
	if err != nil {
		return fmt.Errorf("could not read file '%s': %v", args.file, err)
	}
	
	cleaned := cleanContent(string(content))
	hasXMLDecl := strings.Contains(cleaned, "<?xml")
	
	if !hasXMLDecl {
		fmt.Println("⚠️  XML Best Practice Warnings:")
		fmt.Println("  [XML] Missing XML declaration")
		fmt.Println("    Fix: Add <?xml version=\"1.0\" encoding=\"utf-8\"?> at the top")
		fmt.Println()
		
		if !args.fixWarnings {
			fmt.Println("Use --fix-warnings flag to automatically apply fixes")
			fmt.Println()
		}
	}
	
	// Just process as text to preserve original structure and avoid XML parsing issues
	return processAsText(args, cleaned, hasXMLDecl)
}

func processAsText(args Args, content string, hasXMLDecl bool) error {
	var output bytes.Buffer
	output.Grow(len(content) + 100)
	
	shouldStripXMLDeclaration := false
	if args.fixWarnings && (shouldStripXMLDeclaration || !hasXMLDecl) {
		output.WriteString(XML_DECLARATION)
		if !hasXMLDecl {
			fmt.Println("🔧 Applied fixes:")
			fmt.Println("  ✓ Added XML declaration")
			fmt.Println()
		}
	}
	
	indentLevel := 0
	// Pre-allocate map with estimated size to avoid rehashing
	estimatedElements := len(content) / ESTIMATED_LINE_LENGTH // Estimate based on standard line length
	if estimatedElements < MIN_HASH_CAPACITY {
		estimatedElements = MIN_HASH_CAPACITY
	}
	if estimatedElements > MAX_HASH_CAPACITY {
		estimatedElements = MAX_HASH_CAPACITY
	}
	seenElements := make(map[string]bool, estimatedElements)
	duplicatesRemoved := 0
	
	// Pre-cache common indentation strings (standardized across all implementations)
	indentCache := make([]string, MAX_INDENT_LEVELS+1) // Support up to MAX_INDENT_LEVELS
	for i := 0; i < len(indentCache); i++ {
		indentCache[i] = strings.Repeat("  ", i)
	}
	
	// Process lines with a buffered reader to avoid Scanner token limits
	reader := bufio.NewReader(strings.NewReader(content))
	for {
		line, err := reader.ReadString('\n')
		if len(line) > 0 {
			// Trim only the trailing newline added by ReadString
			if line[len(line)-1] == '\n' {
				line = line[:len(line)-1]
			}
			trimmed := fastTrimSpace(line)
			if trimmed != "" {
				// Never strip XML declaration lines - always preserve them
				// Fast container detection - simple tags without spaces (no attributes)
				isContainer := false
				if len(trimmed) > 2 && trimmed[0] == '<' && trimmed[len(trimmed)-1] == '>' {
					hasSpace := false
					for i := 1; i < len(trimmed)-1; i++ {
						if trimmed[i] == ' ' || trimmed[i] == '\t' {
							hasSpace = true
							break
						}
					}
					isContainer = !hasSpace
				}
				// Deduplication only for non-container lines
				if !isContainer {
					normalizedKey := normalizeWhitespacePreservingAttributes(trimmed)
					if seenElements[normalizedKey] {
						duplicatesRemoved++
						continue // Skip duplicate line - much cleaner than goto
					}
					seenElements[normalizedKey] = true
				}
				// Simplified tag detection
				isClosingTag := len(trimmed) >= 2 && trimmed[0] == '<' && trimmed[1] == '/'
				isOpeningTag := false
				
				// Simple opening tag detection - avoid expensive checks when possible
				if len(trimmed) > 0 && trimmed[0] == '<' && 
				   !(len(trimmed) >= 2 && (trimmed[1] == '/' || trimmed[1] == '!' || trimmed[1] == '?')) &&
				   !strings.Contains(trimmed, "/>") {
					// Only call isSelfContained for potential opening tags
					isOpeningTag = !isSelfContained(trimmed)
				}
				// Adjust indent level for closing tags BEFORE writing the line
				if isClosingTag {
					indentLevel = max(0, indentLevel-1)
				}
				// Apply consistent 2-space indentation using cached strings with optimized writes
				if indentLevel < len(indentCache) {
					output.WriteString(indentCache[indentLevel])
				} else {
					output.WriteString(strings.Repeat("  ", indentLevel)) // Fallback for deep nesting
				}
				output.WriteString(trimmed)
				output.WriteByte('\n')
				// Adjust indent level for opening tags AFTER writing the line
				if isOpeningTag {
					indentLevel++
				}
			}
		}
		if err == io.EOF {
			break
		}
		if err != nil {
			return fmt.Errorf("error reading content: %v", err)
		}
		// Continue to next line if there are more to process
	}
	
	// Ensure final newline for consistency with other implementations  
	if output.Len() > 0 {
		bytes := output.Bytes()
		if bytes[len(bytes)-1] != '\n' {
			output.WriteByte('\n')
		}
	}
	
	outputFilename := getOutputFilename(args.file, args.replace)
	err := os.WriteFile(outputFilename, output.Bytes(), FILE_PERMISSIONS)
	if err != nil {
		return fmt.Errorf("could not write output file: %v", err)
	}
	
	if args.replace {
		err = os.Rename(outputFilename, args.file)
		if err != nil {
			os.Remove(outputFilename)
			return fmt.Errorf("could not replace original file: %v", err)
		}
		fmt.Printf("Original file replaced: %s", args.file)
	} else {
		fmt.Printf("Organized project saved to: %s", outputFilename)
	}
	
	if duplicatesRemoved > 0 {
		fmt.Printf(" (removed %d duplicates)", duplicatesRemoved)
	}
	
	modeText := " (preserving original structure)"
	fmt.Println(modeText)
	
	return nil
}

func getOutputFilename(inputFile string, replaceMode bool) string {
	if replaceMode {
		return inputFile + ".tmp." + strconv.FormatInt(time.Now().Unix(), 10)
	} else {
		if lastDot := strings.LastIndex(inputFile, "."); lastDot != -1 {
			name := inputFile[:lastDot]
			ext := inputFile[lastDot+1:]
			return name + ".organized." + ext
		} else {
			return inputFile + ".organized"
		}
	}
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// fastTrimSpace provides optimized whitespace trimming for XML processing
// Key optimizations:
// - Fast path for strings that don't need trimming (common case)
// - Direct byte comparisons using WHITESPACE_THRESHOLD constant
// - Single allocation when trimming is needed
// Performance: O(1) for pre-trimmed strings, O(n) worst case
func fastTrimSpace(s string) string {
	if len(s) == 0 {
		return s
	}
	
	// Fast path: check endpoints first to avoid scanning (most strings are pre-trimmed)
	if s[0] > WHITESPACE_THRESHOLD && s[len(s)-1] > WHITESPACE_THRESHOLD {
		return s // No allocation needed
	}
	
	// Find first non-whitespace
	start := 0
	for start < len(s) && s[start] <= WHITESPACE_THRESHOLD {
		start++
	}
	
	if start == len(s) {
		return ""
	}
	
	// Find last non-whitespace
	end := len(s) - 1
	for end >= start && s[end] <= WHITESPACE_THRESHOLD {
		end--
	}
	
	if start == 0 && end == len(s)-1 {
		return s
	}
	
	return s[start : end+1]
}

// isSelfContained determines if XML element is complete on single line
// Identifies patterns like <tag>content</tag> to avoid incorrect indentation
// Replaces expensive regex with direct string analysis for better performance
// Performance: O(n) single pass, much faster than regex alternatives
func isSelfContained(s string) bool {
	if len(s) < 7 { // Minimum: <a>b</a>
		return false
	}
	
	// Extract opening tag name
	tagNameStart := 1 // Skip '<'
	tagNameEnd := -1
	
	for i := tagNameStart; i < len(s); i++ {
		if s[i] == ' ' || s[i] == '>' || s[i] == '\t' {
			tagNameEnd = i
			break
		}
	}
	
	if tagNameEnd == -1 {
		return false
	}
	
	tagName := s[tagNameStart:tagNameEnd]
	
	// Check if line ends with </tagname>
	if len(s) >= len(tagName)+3 &&
		s[len(s)-1] == '>' &&
		len(s) >= len(tagName)+3 &&
		s[len(s)-len(tagName)-2] == '/' &&
		s[len(s)-len(tagName)-3] == '<' {
		
		closingTagName := s[len(s)-len(tagName)-2:len(s)-1]
		return len(closingTagName) > 1 && closingTagName[1:] == tagName
	}
	
	return false
}

// normalizeWhitespacePreservingAttributes normalizes structural whitespace while preserving attribute values - optimized
func normalizeWhitespacePreservingAttributes(s string) string {
	if len(s) == 0 {
		return s
	}
	
	// Quick check: if no quotes, use simpler normalization
	if !containsQuotes(s) {
		return normalizeSimpleWhitespace(s)
	}
	
	result := builderPool.Get().(*strings.Builder)
	defer func() {
		result.Reset()
		builderPool.Put(result)
	}()
	result.Grow(len(s))
	
	inQuotes := false
	var quoteChar byte
	prevSpace := false
	expectingAttrValue := false
	
	for i := 0; i < len(s); i++ {
		c := s[i]
		
		if !inQuotes && (c == '"' || c == '\'') {
			inQuotes = true
			quoteChar = c
			expectingAttrValue = false
			// Normalize all quotes to double quotes for consistent deduplication
			result.WriteByte('"')
			prevSpace = false
		} else if inQuotes && c == quoteChar {
			inQuotes = false
			// Normalize all quotes to double quotes for consistent deduplication
			result.WriteByte('"')
			prevSpace = false
		} else if inQuotes {
			// Inside quotes: preserve all whitespace
			result.WriteByte(c)
			prevSpace = false
		} else if c == '=' && !inQuotes {
			// Found attribute assignment, next non-space content might be unquoted value
			result.WriteByte(c)
			expectingAttrValue = true
			prevSpace = false
		} else if expectingAttrValue && c > WHITESPACE_THRESHOLD && c != '>' && c != '/' && c != '"' && c != '\'' {
			// Found unquoted attribute value, add quotes around it
			result.WriteByte('"')
			
			// Find the end of the unquoted value (until space, >, or /)
			j := i
			for j < len(s) && s[j] > WHITESPACE_THRESHOLD && s[j] != '>' && s[j] != '/' {
				result.WriteByte(s[j])
				j++
			}
			result.WriteByte('"')
			i = j - 1 // -1 because the loop will increment
			expectingAttrValue = false
			prevSpace = false
		} else if c <= WHITESPACE_THRESHOLD { // standardized whitespace check
			expectingAttrValue = false
			// Outside quotes: normalize whitespace
			if !prevSpace {
				result.WriteByte(' ')
				prevSpace = true
			}
		} else {
			expectingAttrValue = false
			result.WriteByte(c)
			prevSpace = false
		}
	}
	
	return strings.TrimSpace(result.String())
}

// Helper function to check if string contains quotes - optimized
func containsQuotes(s string) bool {
	return strings.ContainsAny(s, "\"'")
}

// Simplified whitespace normalization for strings without quotes
func normalizeSimpleWhitespace(s string) string {
	// Preserve original whitespace patterns like Zig - only normalize line endings
	// This maintains distinct whitespace variants (tabs vs spaces vs multiple spaces)
	return strings.ReplaceAll(strings.ReplaceAll(s, "\r\n", " "), "\r", " ")
}

func main() {
	args := parseArgs()
	
	if err := processFile(args); err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}
}